File: techinterview/data_structures/trees/tree.c
#include <stdlib.h>
#include <stdio.h>

// BST node data type
struct node
{
    int value;
    struct node* left;
    struct node* right;
};

// node functions
struct node* node_new(int value)
{
    struct node* n = malloc(sizeof(struct node));
    n->value = value;
    n->left = NULL;
    n->right = NULL;
    return n;
}

void node_destroy(struct node* n)
{
    if(n)
    {
        node_destroy(n->left);
        node_destroy(n->right);
        free(n);
    }
}

struct node* node_min_node_in_subtree(struct node* n)
{
    struct node* tmp = n;

    if(tmp)
    {
        while(tmp->left != NULL)
        {
            tmp = tmp->left;
        }
    }

    return tmp;
}

void node_insert(struct node* n, int value)
{
    if(value < n->value)
    {
        if(n->left == NULL)
        {
            n->left = node_new(value);
            return;
        }
        else
        {
            node_insert(n->left, value);
        }
    }
    else
    {
        if(n->right == NULL)
        {
            n->right = node_new(value);
            return;
        }
        else
        {
            node_insert(n->right, value);
        }
    }
}

struct node* node_delete(struct node* n, int value)
{
    if(n == NULL)
    {
        return NULL;
    }

    // search in left subtree
    if(value < n->value)
    {
        n->left = node_delete(n->left, value);
    }
    // search in righ subtree
    else if(value > n->value)
    {
        n->right = node_delete(n->right, value);
    }
    // found it
    else
    {
        struct node* tmp;

        // one child, promote right
        if(n->left == NULL)
        {
            tmp = n->right;
            free(n);
            return tmp;
        }
        // one child, promote left
        if(n->right == NULL)
        {
            tmp = n->left;
            free(n);
            return tmp;
        }

        // two children, copy minimum to this node and delete the original minimum
        // find minimum value in right subtree
        tmp = node_min_node_in_subtree(n->right);
        // replace this node with it's value
        n->value = tmp->value;
        // delete the original which is now a duplicate
        n->right = node_delete(n->right, tmp->value);
    }

    return n;
}

void node_print_inorder(struct node* n)
{
    if(n)
    {
        node_print_inorder(n->left);
        fprintf(stderr, "Node(%d)\n", n->value);
        node_print_inorder(n->right);
    }
}

// BST tree data type
struct tree
{
    struct node* root;
};

// BST tree functions
struct tree* new_tree()
{
    return malloc(sizeof(struct tree));
}

void tree_destroy(struct tree* t)
{
    node_destroy(t->root);
    free(t);
}

void tree_insert(struct tree* t, int value)
{
    if(t->root == NULL)
    {
        t->root = malloc(sizeof(struct node));
        t->root->left = NULL;
        t->root->right = NULL;
        t->root->value = value;
    }
    else
    {
        node_insert(t->root, value);
    }
}

void tree_delete(struct tree* t, int value)
{
    node_delete(t->root, value);
}

void tree_print_inorder(struct tree* t)
{
    fprintf(stderr, "Tree:\n");
    node_print_inorder(t->root);
}


int main(int argc, char* argv[])
{
    // create new tree
    struct tree *t = new_tree();

    // insert a bunch of nodes
    tree_insert(t, 50);
    tree_insert(t, 30);
    tree_insert(t, 20);
    tree_insert(t, 40);
    tree_insert(t, 70);
    tree_insert(t, 60);
    tree_insert(t, 80);

    // tree:
    //         50
    //     30      70
    //  20   40  60   80


    // 20->30->40->50->60->70->80
    tree_print_inorder(t);

    tree_delete(t, 20);
    // tree:
    //         50
    //     30      70
    //       40  60   80

    // 30->40->50->60->70->80
    tree_print_inorder(t);

    tree_delete(t, 30);
    // tree:
    //         50
    //     40      70
    //           60   80

    // 40->50->60->70->80
    tree_print_inorder(t);

    tree_delete(t, 50);
    // tree:
    //     60
    // 40      70
    //            80

    // 40->60->70->80
    tree_print_inorder(t);

    tree_destroy(t);

    return 0;
}

File: techinterview/data_structures/trees/go/tree.go
package main

import "fmt"

type Node struct {
    Right *Node
    Left *Node
    Value int
}

type Tree struct {
    Root *Node
}

// create a new node with no children and value set to value
func newNode(value int) *Node{
    return &Node{nil, nil, value}
}

// recursively search through subtrees to find location to insert new node
func (n *Node) insert(value int) {
    if value < n.Value {
        if n.Left == nil {
            n.Left = newNode(value)
        } else {
            n.Left.insert(value)
        }
    } else{
        if n.Right == nil {
            n.Right = newNode(value)
        } else {
            n.Right.insert(value)
        }
    }
}

// recursively traverse a tree in-order printing each node
func (n *Node) traverse() {
    if n != nil {
        n.Left.traverse()
        fmt.Printf("Node(%d)\n", n.Value)
        n.Right.traverse()
    }
}

// recursively traverses a tree search for a node that has the value. return true if it is found, false otherwise
func (n *Node) findValue(value int) bool {
    if n == nil {
        return false
    }
    if value == n.Value {
        return true
    }
    if value < n.Value {
        return n.Left.findValue(value)
    } else {
        return n.Right.findValue(value)
    }
}

// find the minimum node within a node's subtree
func (n *Node) findMinNodeInSubtree() *Node {
    if n.Left == nil {
        return n
    }
    return n.Left.findMinNodeInSubtree()
}

// recursively traverse a node's subtree and delete a node matching value
func (n *Node) delete(value int) *Node {
    if n == nil {
        return nil
    }
    if value < n.Value {
        n.Left = n.Left.delete(value)
    } else if value > n.Value {
        n.Right = n.Right.delete(value)
    } else {
        if n.Left == nil {
            return n.Right
        }
        if n.Right == nil {
            return n.Left
        }
        // two children, find the min and swap with this value
        minnode := n.Right.findMinNodeInSubtree()
        n.Value = minnode.Value
        n.Right = n.Right.delete(minnode.Value)
    }

    return n
}

// return a pointer to a new tree with no root
func NewTree() *Tree{
    return &Tree{nil}
}

// insert a node with value == value into a tree
func (t *Tree) Insert(value int) {
    if t.Root == nil{
        t.Root = newNode(value)
    } else {
        t.Root.insert(value)
    }
}

// searches for a node whose value is value and returns a bool indicating if it is present in the tree
func (t *Tree) Search(value int) bool {
    if t.Root == nil {
        return false
    }
    return t.Root.findValue(value)
}

func (t *Tree) Delete(value int) {
    if t.Root != nil {
        t.Root.delete(value)
    }
}

// traverse a tree in-order and print a node on each newline
func (t *Tree) Traverse() {
    fmt.Println("Tree:")
    t.Root.traverse()
}

func main() {
    t := NewTree()
    t.Insert(50);
    t.Insert(30);
    t.Insert(20);
    t.Insert(40);
    t.Insert(70);
    t.Insert(60);
    t.Insert(80);

    // tree:
    //         50
    //     30      70
    //  20   40  60   80


    // 20->30->40->50->60->70->80
    t.Traverse();

    t.Delete(20);
    // tree:
    //         50
    //     30      70
    //       40  60   80

    // 30->40->50->60->70->80
    t.Traverse();

    t.Delete(30);
    // tree:
    //         50
    //     40      70
    //           60   80

    // 40->50->60->70->80
    t.Traverse();

    t.Delete(50);
    // tree:
    //     60
    // 40      70
    //            80

    // 40->60->70->80
    t.Traverse();
}
